src/transformStateWithClones.js

```diff
@@ -7,7 +7,39 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    let newState;
+
+    switch (action.type) {
+      case 'clear':
+        newState = {};
+        break;
+
+      case 'addProperties':
+        newState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        newState = { ...currentState };
+
+        action.keysToRemove.forEach((key) => {
+          delete newState[key];
+        });
+        break;
+      default:
+        throw new Error(`Unknown action type: ${action.type}`);
+    }
+
+    currentState = newState;
+
+    stateHistory.push(newState);
```

Comment:
It's good practice to create a new object here before pushing to `stateHistory` to ensure that each entry in the history array is a distinct object. However, this can be done in a more concise way by using the spread operator inside the push method like so: `stateHistory.push({ ...currentState });`